playground/heat_equation_cn.py

```diff
@@ -0,0 +1,130 @@
+import numpy as np
+from scipy.sparse.linalg import spsolve
+from scipy.sparse import csc_matrix
+from matplotlib import pyplot as plt
+
+# setting up parameters
+length = 1
+nodes = 100
+time = 30
+k = 0.01
+initial_temp = 25
+
+
+def left_temp_func(temp, t):
+    """
+    Generates the temperature at the left boundary of the rod
+
+    :param temp: initial temperature of the rod at t=0
+    :param t: array consisting of time steps from t=0 to t= total time
+    :return: the temperature at the left boundary
+    """
+    return 20 * np.sin(np.pi * t) + temp
+
+
+def right_temp_func(temp, t):
+    """
+    Generates the temperature at the right boundary of the rod
+
+    :param temp: initial temperature of the rod at t=0
+    :param t: array consisting of time steps from t=0 to t= total time
+    :return: the temperature at the right boundary
+    """
+    return t + temp
+
+
+def build_tridiagonal_matrix(a, b, c, nodes):
+    """
+    Initialises the tridiagonal matrix on the LHS of the equation
+
+    :param a: the coefficient of U @ (t = tau + 1 & x = i-1)
+    :param b: the coefficient of U @ (t = tau + 1 & x = i)
+    :param c: the coefficient of U @ (t = tau + 1 & x = i+1)
+    :param nodes: number of spatial nodes ( used to initialise the size of the tridiagonal matrix)
+    :return: the tridiagonal matrix consisting of coefficients
+    """
+
+    matrix = np.zeros((nodes, nodes))
+    np.fill_diagonal(matrix, b)
+    np.fill_diagonal(matrix[1:], a)
+    np.fill_diagonal(matrix[:, 1:], c)
+
+    matrix = csc_matrix(matrix)
+
+    return matrix
+
+
+def solve_heat_equation_cn(length, nodes, time, k, initial_temp):
+    """
+    Solves the heat equation using the crank nicolson method
+
+    :param length: length of the rod
+    :param nodes: number of spatial nodes across the rod
+    :param time: the total time
+    :param k: heat diffusivity constant
+    :param initial_temp: initial temperature of the rod at t=0
+    :return: - u: the solved 2d matrix consisting the temperature distribution over time
+             - dt: time step size (used for plotting)
+             - time_step: no. of time steps for the simulation (used for plotting)
+    """
+    dx = length / (nodes-1)
```

Comment:
Lines 71-72 are unnecessary. The Crank-Nicolson method is **unconditionally stable**.